account_invoice_triple_discount/models/account_invoice.py

```diff
@@ -54,6 +54,9 @@ def triple_discount_preprocess(self):
         Resetting discount2 and discount3 to 0.0 avoids issues if
         this method is called multiple times."""
         prev_values = dict()
+        if not self.env.in_onchange:
```

Comment:
@SimoRubi Can you elaborate why this trick is required?

Reply:
@lmignon I don't know exactly why this works, I expected all the changes of #500 to work without any cache management.

The really wrong thing in my opinion is that the automated tests do not work without the `invalidate_cache` trick.
While debugging the tests issue (that note, 1 in 5 times were not failing!) I noticed the following (extracted from this same file):
```
    @api.multi
    @api.depends('discount2', 'discount3')
    def _compute_price(self):
        prev_values = self.triple_discount_preprocess()
        # Here discount is 50 and discount3 is 0, but inside the following method discount is 0 again.
        # More in particular, the record's discounts values change after a _browse call in the model's __iter__ method
        super(AccountInvoiceLine, self)._compute_price()
        self.triple_discount_postprocess(prev_values)
```
Moreover, this happened only for the 2nd of the lines created in `test_04`..
So I concluded that something was really messed up with the cache management

So I had to add the `invalidate_cache` line, but that somehow breaks the interface behavior as explained in https://github.com/OCA/account-invoicing/pull/517/commits/4411c63c9c2b807f8c17b8354e1ce85085a50f2a..

Reply:
@SimoRubi IMO it's not the right fix. If something is wrong with the cache we must fix it into Odoo. This error could  also reveal a problem into the way the 2 addons are designed.

Reply:
@lmignon I agree this is not the right fix, I would simply call this a workaround.
Anyway I probably won't have time to investigate further into the Odoo cache management and find the right fix (for these modules, or for the Odoo core).